loadtest/control/simulcontroller/actions.go

```diff
@@ -25,22 +26,45 @@ type userAction struct {
 }
 
 func (c *SimulController) connect() error {
+	if !atomic.CompareAndSwapInt32(&c.connectedFlag, 0, 1) {
+		return errors.New("already connected")
+	}
 	errChan, err := c.user.Connect()
 	if err != nil {
+		atomic.StoreInt32(&c.connectedFlag, 0)
 		return fmt.Errorf("connect failed %w", err)
 	}
+	c.wg.Add(3)
 	go func() {
+		defer c.wg.Done()
 		for err := range errChan {
 			c.status <- c.newErrorStatus(err)
 		}
 	}()
+	go c.wsEventHandler(c.wg)
+	go c.periodicActions(c.wg)
```

Comment:
Any reason we are passing the waitgroup? We have it as part of the controller after all.

Reply:
Just to make more explicit where `wg` is incremented and consumed.

Reply:
Fair enough :+1: